test_conformance/workgroups/test_wg_scan_reduce.cpp

```diff
@@ -112,29 +163,133 @@ template <typename C> struct Reduce
 
             for (size_t j = 0; j < wg_size; j++)
             {
-                if (result != outptr[i + j])
+                if constexpr (std::is_floating_point_v<Type>)
                 {
-                    log_info("%s_%s: Error at %zu\n", testName, testOpName,
-                             i + j);
-                    return -1;
+                    if (fabs(result - outptr[i + j]) > max_err)
+                    {
+                        log_info("%s_%s: Error at %zu\n", testName, testOpName,
+                                 i + j);
+                        return -1;
+                    }
+                }
+                else if (std::is_same_v<Type, cl_half>)
+                {
+                    if (fabs(cl_half_to_float(result)
+                             - cl_half_to_float(outptr[i + j]))
+                        > cl_half_to_float(max_err))
+                    {
+                        log_info("%s_%s: Error at %zu\n", testName, testOpName,
+                                 i + j);
+                        return -1;
+                    }
+                }
+                else
+                {
+                    if (result != outptr[i + j])
+                    {
+                        log_info("%s_%s: Error at %zu\n", testName, testOpName,
+                                 i + j);
+                        return -1;
+                    }
                 }
             }
         }
         return 0;
     }
+
+    static void generate_reference_values(Type *inptr, size_t n_elems,
+                                          size_t max_wg_size, Type &max_err = 0)
+    {
+        MTdataHolder d(gRandomSeed);
+        if constexpr (std::is_floating_point_v<
+                          Type> || std::is_same_v<Type, cl_half>)
+        {
+            std::vector<Type> ref_vals(max_wg_size, 0);
+            if (std::is_same_v<Type, cl_half>)
+            {
+                // to prevent overflow limit range of randomization
+                float max_range = 99.0;
+                float min_range = -99.0;
+                // generate reference values for one work group
```

Comment:
Do we want at least a couple of work-groups to test special values, things like NaNs, infinities, maybe +/-0, etc?

Reply:
Sure, do you think we can go for new test just to keep special cases separated ? I presume it may be more readable.